### DISCUSSION TITLE
dynamic export type

### DISCUSSION THREAD

**Discussion by CodeLazier** [godotengine/godot-proposals#7016] [Editor]
Is there a way to dynamically change the type of @export var,
Ex:
I have an enum:
```gdscript
enum c{food,tools}
```
There is a subcategory:
```gdscript
enum sc_food{apple, banana}
enum sc_tool{axe,saw}
```

my idea is that you can then select categories and so on:
```gdscript
@export var category:c
@export var sub_category:int
```

because dynamic type conversion is not supported, sub_category can only be defined as int.
If something like this:
```gdscript
@export var category:c:
set(v):
    if category != v:
      category = v
      match v:
        c.food:
          _change_export_type(self,sub_category,sc_food)
        c.tool:
          _change_export_type(self,sub_category,sc_tool)
```
This will then display the name I want to choose in the form of enum in the inspector

**Comment by AThousandShips** [comment]
You can achieve this with `_get_property_list`, but not sure how advisable it is, and how well it will work for loading, see for example godotengine/godot#77259

**Comment by dalexeev** [comment]
See godotengine/godot#75778. If/when `_validate_property` becomes available for scripting, you will be able to do that and other things.

**Comment by neth392** [comment]
I came across this issue myself and wrote the resource class below to allow me to have an exported var that has a type not specified in code, but in the editor. It's not perfect or the cleanest in the editor, but it works for my case. I just used an Array and limited it to 1 element since Arrays support any type of Variant. I also make sure that the variant matches the specified type because I want to be sure the type of the value stays consistent for each resource instance.

```
@tool
class_name GameSetting extends Resource

@export var id: StringName
@export var type: Variant.Type:
	set(_type):
		if _type == null || typeof(value[0]) != _type:
			value[0] = null
		type = _type
@export var value: Array = [null]:
	set(_value):
		if _value == null || _value.size() == 0: # Make sure the array exists & has 1 element
			value = [null]
			push_warning("Warning: GameSettings.value must be an Array with size of 1")
			return
		
		if _value.size() > 1: # Make sure the array doesn't have more than 1 element
			value = [_value[0]]
			push_warning("Warning: GameSettings.value must be an Array with size of 1")
			return
		
		if typeof(_value[0]) != type: # Make sure the type matches the expected type
			push_warning("Warning: GameSettings.value type must match GameSettings.type")
			return
		
		value = _value
```

**Comment by Jack-023** [comment]
I think that we should be able to export Variants. eg
```gdscript
@export var foo: Variant
```
It could use the same interface in the inspector as untyped Array values or Dictionary keys/values. Currently this can be worked around by just exporting an array and accessing the first value as mentioned above but that feels a bit hacky when there is already a way to set Variant data in the inspector.

**Reply by Mickeon** [reply]
- See https://github.com/godotengine/godot/pull/89324
